### PR TITLE
fix(Swipe-to-Receive): move and finish implementation of check for un…

### PR DESCRIPTION
**Pull request by woudini**
…used addresses

### CODE REVIEW

File: Blockchain/Addresses/AssetAddressRepository.swift [hunk 1]
```diff
@@ -66,7 +66,7 @@ import Foundation
     ///
     /// - Parameter assetType: the AssetType
     /// - Returns: the swipe addresses
-    @objc func swipeToReceiveAddresses(for assetType: AssetType) -> [String] {
+    @objc func swipeToReceiveAddresses(for assetType: AssetType) -> [Any] {
```
Comment (bchrisarriola) [inline Blockchain/Addresses/AssetAddressRepository.swift]: Why was this changed to `Any`? I think this should be reverted to `String` since addresses won't ever be any other type.

File: Blockchain/Addresses/AssetAddressRepository.swift [hunk 2]
```diff
@@ -99,9 +99,51 @@ import Foundation
...
         }
     }
 }
+
+extension AssetAddressRepository {
+    @objc func checkForUnusedAddress(_ address: String,
+                                     displayAddress: String,
+                               legacyAssetType: LegacyAssetType,
+                               successHandler: @escaping ((_ address: String, _ isUnused: Bool) -> Void),
+                               errorHandler: @escaping (() -> Void)) {
```
Comment (bchrisarriola) [inline Blockchain/Addresses/AssetAddressRepository.swift]: Alignment is a bit weird here. I think this structure is much more readable:

```
@objc func checkForUnusedAddress(
    _ address: String,
    displayAddress: String,
    legacyAssetType: LegacyAssetType,
    successHandler: @escaping ((_ address: String, _ isUnused: Bool) -> Void),
    errorHandler: @escaping (() -> Void)
) {
// ...
}
```

File: PinEntry/Classes/PEPinEntryController.m [hunk 3]
```diff
@@ -222,7 +222,19 @@ - (void)addAddressToSwipeView:(BCSwipeAddressView *)swipeView assetType:(LegacyA
...
                 }
             };
-            [app checkForUnusedAddress:nextAddress success:success error:error assetType:assetType];
+            
+            NSString *addressToCheck = nextAddress;
+            NSString *displayAddress = nextAddress;
+
+            if (assetType == LegacyAssetTypeBitcoinCash) {
+                if ([nextAddress isKindOfClass:[NSString class]]) {
+                    [KeychainItemWrapper removeAllSwipeAddressesForAssetType:LegacyAssetTypeBitcoinCash];
```
Comment (bchrisarriola) [inline PinEntry/Classes/PEPinEntryController.m]: Use `AssetAddressRepository`. UIViewControllers shouldn't interface with low-level constructs like the keychain/nsuserdefaults etc.

File: Blockchain/Addresses/AssetAddressRepository.swift [hunk 4]
```diff
@@ -99,9 +99,51 @@ import Foundation
...
-    func onRetrievedSwipeToReceive(addresses: [String], assetType: AssetType) {
+    func onRetrievedSwipeToReceive(addresses: [Any], assetType: AssetType) {
         addresses.forEach {
             KeychainItemWrapper.addSwipeAddress($0, assetType: assetType.legacy)
         }
     }
 }
+
+extension AssetAddressRepository {
+    @objc func checkForUnusedAddress(_ address: String,
```
Comment (bchrisarriola) [inline Blockchain/Addresses/AssetAddressRepository.swift]: This used to accept `AssetAddress`, any reason that was changed?

File: Blockchain/Addresses/AssetAddressRepository.swift (hunk 1 above)
Comment (woudini) [reply Blockchain/Addresses/AssetAddressRepository.swift]: Sorry I should have clarified in the PR description or added a comment and asked for feedback. Currently in the `swift` branch BCH addresses are stored in the BCH format `bitcoincash:address`. However the BCH lookup URL for `checkForUnusedAddress` requires the corresponding BTC address. Before this refactor,`checkForUnusedAddress` was calling `[app.wallet fromBitcoinCash:address]`. Here however the `WalletManager` singleton or a part of it is not yet initialized at that point, so I chose to do the following:
1. store a dictionary of both the BCH address and corresponding BTC address when retrieving the addresses
2. if the current BCH address is not in dictionary format, remove swipe addresses for BCH

I did this under the impression that we were on schedule to get a build done today.

File: Blockchain/Addresses/AssetAddressRepository.swift (hunk 4 above)
Comment (woudini) [reply Blockchain/Addresses/AssetAddressRepository.swift]: I can't create an `AssetAddress` in obj-c

Comment (bchrisarriola) [reply Blockchain/Addresses/AssetAddressRepository.swift]: For it to be accessible in objective-c, you'll need to add the `@objc` annotation to the class.

i.e. 
```
@objc public protocol AssetAddress
```

**Comment by bchrisarriola** [comment]
It's generally not a good pattern to return very generic objects (such as `id`). This is why we've been refactoring NSDictionary references to very specific concrete objects (i.e. `AssetAddress`, `BitcoinUrlPayload`, etc.). It's easier to work with typed objects since you know what contract you can use/expect.

Since we're mostly trying to finish `RootService` deprecation, can we break this PR up in 2? (1) migrate to use the existing `checkForUnusedAddress` and (2) implement BTC/BCH

**Comment by bchrisarriola** [comment]
What's the status with this PR? Should we close this if you're creating a new one or do you plan on updating it?

**Comment by woudini** [comment]
I'll do a new one since #199 has some of these changes
